src/test/java/com/redhat/devtools/lsp4ij/features/formatting/TypeScriptEditorImprovementsTest.java

```diff
@@ -0,0 +1,509 @@
+        assertEquals("console.log(" + outerQuote + outerQuote + ");", document.getText());
+        assertEquals(initialOffset + 2, caretModel.getOffset());
+
+        // Return to the interior of the string literal and type an inner quote character which should be inserted unpaired
+        caretModel.moveToOffset(initialOffset + 1);
+        EditorTestUtil.performTypingAction(editor, innerQuote);
+        assertEquals("console.log(" + outerQuote + innerQuote + outerQuote + ");", document.getText());
+        assertEquals(initialOffset + 2, caretModel.getOffset());
+
+        // Typing another inner quote character should also result in unpaired insertion
+        EditorTestUtil.performTypingAction(editor, innerQuote);
+        assertEquals("console.log(" + outerQuote + innerQuote + innerQuote + outerQuote + ");", document.getText());
+        assertEquals(initialOffset + 3, caretModel.getOffset());
+
+        // Typing a backspace should remove the second inner quote only
+        EditorTestUtil.executeAction(editor, IdeActions.ACTION_EDITOR_BACKSPACE);
+        assertEquals("console.log(" + outerQuote + innerQuote + outerQuote + ");", document.getText());
+        assertEquals(initialOffset + 2, caretModel.getOffset());
+
+        // And another should leave us with the empty outer quote
+        EditorTestUtil.executeAction(editor, IdeActions.ACTION_EDITOR_BACKSPACE);
+        assertEquals("console.log(" + outerQuote + outerQuote + ");", document.getText());
+        assertEquals(initialOffset + 1, caretModel.getOffset());
+
+        // Now insert an escaped outer quote which should also not be paired
+        EditorTestUtil.performTypingAction(editor, BACKSLASH);
+        EditorTestUtil.performTypingAction(editor, outerQuote);
+        assertEquals("console.log(" + outerQuote + BACKSLASH + outerQuote + outerQuote + ");", document.getText());
+        assertEquals(initialOffset + 3, caretModel.getOffset());
+
+        // Remove those and make sure that nothing else is removed
+        EditorTestUtil.executeAction(editor, IdeActions.ACTION_EDITOR_BACKSPACE);
+        assertEquals("console.log(" + outerQuote + BACKSLASH + outerQuote + ");", document.getText());
+        assertEquals(initialOffset + 2, caretModel.getOffset());
+        EditorTestUtil.executeAction(editor, IdeActions.ACTION_EDITOR_BACKSPACE);
+        assertEquals("console.log(" + outerQuote + outerQuote + ");", document.getText());
+        assertEquals(initialOffset + 1, caretModel.getOffset());
+
+        // And finally remove the original outer quote and the quote pair should be removed
+        EditorTestUtil.executeAction(editor, IdeActions.ACTION_EDITOR_BACKSPACE);
+        assertEquals("console.log();", document.getText());
+        assertEquals(initialOffset, caretModel.getOffset());
+    }
+
+    // Confirms the default/"broken" behavior when 'enableStringLiteralImprovements' is disabled
+    private void testNestedQuotesDisabled(char outerQuote, char innerQuote) {
```

Comment:
The disabled tests just verify that things behave as they would -- IMHO, improperly -- without these improvements and fixes.

And that'll be the common pattern as you go through this test class...